Bottom: 68c8a49f409c193cd7d4f56c400950af49a32821
Top:    c138bafea9277bd0a89ab68238b396f370d335f4
Author: Jon Smirl <jonsmirl@gmail.com>
Date:   2009-05-09 13:56:07 -0400

Refresh of stac9766-codec

---

diff --git a/sound/soc/codecs/stac9766.c b/sound/soc/codecs/stac9766.c
index b32b7bf..d7996ff 100644
--- a/sound/soc/codecs/stac9766.c
+++ b/sound/soc/codecs/stac9766.c
@@ -23,6 +23,7 @@
 #include <sound/initval.h>
 #include <sound/pcm_params.h>
 #include <sound/soc.h>
+#include <sound/tlv.h>
 #include <sound/soc-of-simple.h>
 
 #include "stac9766.h"
@@ -80,40 +81,45 @@ static const struct soc_enum stac9766_boost2_enum =
 static const struct soc_enum stac9766_stereo_mic_enum =
 	SOC_ENUM_SINGLE(AC97_STAC_STEREO_MIC, 2, 1, stac9766_stereo_mic);
 
+static const DECLARE_TLV_DB_LINEAR(master_tlv, -4600, 0);
+static const DECLARE_TLV_DB_LINEAR(record_tlv, 0, 2250);
+static const DECLARE_TLV_DB_LINEAR(beep_tlv, -4500, 0);
+static const DECLARE_TLV_DB_LINEAR(mix_tlv, -3450, 1200);
+
 static const struct snd_kcontrol_new stac9766_snd_ac97_controls[] = {
-	SOC_DOUBLE("Speaker Volume", AC97_MASTER, 8, 0, 31, 1),
+	SOC_DOUBLE_TLV("Speaker Volume", AC97_MASTER, 8, 0, 31, 1, master_tlv),
 	SOC_SINGLE("Speaker Switch", AC97_MASTER, 15, 1, 1),
-	SOC_DOUBLE("Headphone Volume", AC97_HEADPHONE, 8, 0, 31, 1),
+	SOC_DOUBLE_TLV("Headphone Volume", AC97_HEADPHONE, 8, 0, 31, 1, master_tlv),
 	SOC_SINGLE("Headphone Switch", AC97_HEADPHONE, 15, 1, 1),
-	SOC_SINGLE("Mono Out Volume", AC97_MASTER_MONO, 0, 31, 1),
+	SOC_SINGLE_TLV("Mono Out Volume", AC97_MASTER_MONO, 0, 31, 1, master_tlv),
 	SOC_SINGLE("Mono Out Switch", AC97_MASTER_MONO, 15, 1, 1),
 
-	SOC_DOUBLE("Record Volume", AC97_REC_GAIN, 8, 0, 15, 1),
+	SOC_DOUBLE_TLV("Record Volume", AC97_REC_GAIN, 8, 0, 15, 1, record_tlv),
 	SOC_SINGLE("Record Switch", AC97_REC_GAIN, 15, 1, 1),
 
 
-	SOC_SINGLE("Beep Volume", AC97_PC_BEEP, 1, 15, 1),
+	SOC_SINGLE_TLV("Beep Volume", AC97_PC_BEEP, 1, 15, 1, beep_tlv),
 	SOC_SINGLE("Beep Switch", AC97_PC_BEEP, 15, 1, 1),
 	SOC_SINGLE("Beep Frequency", AC97_PC_BEEP, 5, 127, 1),
-	SOC_SINGLE("Phone Volume", AC97_PHONE, 0, 31, 1),
+	SOC_SINGLE_TLV("Phone Volume", AC97_PHONE, 0, 31, 1, mix_tlv),
 	SOC_SINGLE("Phone Switch", AC97_PHONE, 15, 1, 1),
 
 	SOC_ENUM("Mic Boost1", stac9766_boost1_enum),
 	SOC_ENUM("Mic Boost2", stac9766_boost2_enum),
-	SOC_SINGLE("Mic Volume", AC97_MIC, 0, 31, 1),
+	SOC_SINGLE_TLV("Mic Volume", AC97_MIC, 0, 31, 1, mix_tlv),
 	SOC_SINGLE("Mic Switch", AC97_MIC, 15, 1, 1),
 	SOC_ENUM("Stereo Mic", stac9766_stereo_mic_enum),
 
-	SOC_DOUBLE("Line Volume", AC97_LINE, 8, 0, 31, 1),
+	SOC_DOUBLE_TLV("Line Volume", AC97_LINE, 8, 0, 31, 1, mix_tlv),
 	SOC_SINGLE("Line Switch", AC97_LINE, 15, 1, 1),
-	SOC_DOUBLE("CD Volume", AC97_CD, 8, 0, 31, 1),
+	SOC_DOUBLE_TLV("CD Volume", AC97_CD, 8, 0, 31, 1, mix_tlv),
 	SOC_SINGLE("CD Switch", AC97_CD, 15, 1, 1),
-	SOC_DOUBLE("AUX Volume", AC97_AUX, 8, 0, 31, 1),
+	SOC_DOUBLE_TLV("AUX Volume", AC97_AUX, 8, 0, 31, 1, mix_tlv),
 	SOC_SINGLE("AUX Switch", AC97_AUX, 15, 1, 1),
-	SOC_DOUBLE("Video Volume", AC97_VIDEO, 8, 0, 31, 1),
+	SOC_DOUBLE_TLV("Video Volume", AC97_VIDEO, 8, 0, 31, 1, mix_tlv),
 	SOC_SINGLE("Video Switch", AC97_VIDEO, 15, 1, 1),
 
-	SOC_DOUBLE("DAC Volume", AC97_PCM, 8, 0, 31, 1),
+	SOC_DOUBLE_TLV("DAC Volume", AC97_PCM, 8, 0, 31, 1, mix_tlv),
 	SOC_SINGLE("DAC Switch", AC97_PCM, 15, 1, 1),
 	SOC_SINGLE("Loopback Test Switch", AC97_GENERAL_PURPOSE, 7, 1, 0),
 	SOC_SINGLE("3D Volume", AC97_3D_CONTROL, 3, 2, 1),
